File: training-data/ocaml/kosaraju_scc.ml
let build_reverse graph n =
  let rev = Array.make n [] in
  Array.iteri (fun u neighbors -> List.iter (fun v -> rev.(v) <- u :: rev.(v)) neighbors) graph;
  rev

let kosaraju_scc graph n =
  let visited = Array.make n false in
  let order = ref [] in
  let rec dfs1 u =
    visited.(u) <- true;
    List.iter (fun v -> if not visited.(v) then dfs1 v) graph.(u);
    order := u :: !order
  in
  for u = 0 to n - 1 do
    if not visited.(u) then dfs1 u
  done;

  let rev = build_reverse graph n in
  let assigned = Array.make n false in
  let components = ref [] in
  let rec dfs2 u component =
    assigned.(u) <- true;
    component := u :: !component;
    List.iter (fun v -> if not assigned.(v) then dfs2 v component) rev.(u)
  in
  List.iter
    (fun u ->
      if not assigned.(u) then begin
        let component = ref [] in
        dfs2 u component;
        components := !component :: !components
      end)
    !order;
  !components

let () =
  (* 0 -> 1 -> 2 -> 0 (cycle), 2 -> 3 -> 4 -> 3 (cycle) *)
  let graph = [| [ 1 ]; [ 2 ]; [ 0; 3 ]; [ 4 ]; [ 3 ] |] in
  let sccs = kosaraju_scc graph 5 in
  List.iter
    (fun comp ->
      List.iter (fun v -> Printf.printf "%d " v) comp;
      print_newline ())
    sccs

File: training-data/python/kosaraju_scc.py
"""Kosaraju's algorithm: two depth-first passes find strongly connected
components in a directed graph -- one over the graph in finish-time
order, one over its transpose, grouping nodes that can reach each
other both ways. (tarjan_scc.py solves the same problem with a single
pass and a low-link stack; this is the two-pass alternative.)"""

from collections import defaultdict
from typing import Dict, List


def kosaraju_scc(graph: Dict[str, List[str]]) -> List[List[str]]:
    visited = set()
    finish_order = []

    def dfs1(node):
        stack = [(node, iter(graph.get(node, [])))]
        visited.add(node)
        while stack:
            current, neighbors = stack[-1]
            advanced = False
            for neighbor in neighbors:
                if neighbor not in visited:
                    visited.add(neighbor)
                    stack.append((neighbor, iter(graph.get(neighbor, []))))
                    advanced = True
                    break
            if not advanced:
                finish_order.append(current)
                stack.pop()

    for node in graph:
        if node not in visited:
            dfs1(node)

    reverse_graph = defaultdict(list)
    for node, neighbors in graph.items():
        for neighbor in neighbors:
            reverse_graph[neighbor].append(node)

    visited.clear()
    components = []

    def dfs2(node, component):
        stack = [node]
        visited.add(node)
        while stack:
            current = stack.pop()
            component.append(current)
            for neighbor in reverse_graph.get(current, []):
                if neighbor not in visited:
                    visited.add(neighbor)
                    stack.append(neighbor)

    for node in reversed(finish_order):
        if node not in visited:
            component: List[str] = []
            dfs2(node, component)
            components.append(component)

    return components


if __name__ == "__main__":
    graph = {
        "a": ["b"],
        "b": ["c"],
        "c": ["a", "d"],
        "d": ["e"],
        "e": ["f"],
        "f": ["d"],
    }
    for scc in kosaraju_scc(graph):
        print(sorted(scc))
